File: src/ludoxel/foundations/mathematics/voxels/dda.py
# SPDX-FileCopyrightText: 2026 Kento Konishi
# SPDX-License-Identifier: LicenseRef-All-Rights-Reserved
from __future__ import annotations

import math
from dataclasses import dataclass

from ludoxel.foundations.mathematics.linear.vec3 import Vec3


@dataclass(frozen=True)
class DDAHit:
  """
  voxel grid traversal の各 step で訪問された cell と ray parameter を保持する。
  `cell_x`、`cell_y`、`cell_z` は grid cell index、`t` は `origin + direction * t` の parameter、
  `enter_face` は直前境界を越えて現在 cell に入った face index 又は開始 cell を表す -1 である。
  """

  cell_x: int
  cell_y: int
  cell_z: int
  t: float
  enter_face: int = -1


def dda_grid_traverse(origin: Vec3, direction: Vec3, t_max: float, cell_size: float = 1.0):
  """
  ray が通過する voxel cell を Amanatides-Woo 型の DDA で順に生成する。
  `origin` と `direction` は同じ world 座標系の `Vec3`、`cell_size` は正の grid 間隔として扱われ、direction が零 vector の場合は何も yield しない。
  この Python source は native extension build の対象でもあり、block picking と third-person camera collision は traversal 順序と face index の同じ契約に依存する。
  """
  d = direction
  if abs(d.x) < 1e-12 and abs(d.y) < 1e-12 and abs(d.z) < 1e-12:
    return

  x = math.floor(origin.x / cell_size)
  y = math.floor(origin.y / cell_size)
  z = math.floor(origin.z / cell_size)

  step_x = 1 if d.x > 0 else -1
  step_y = 1 if d.y > 0 else -1
  step_z = 1 if d.z > 0 else -1

  def int_bound(s: float, ds: float) -> float:
    """
    現在位置から次の整数 grid 境界までの ray parameter 増分を一軸について計算する。
    `ds > 0` では上側境界までの距離、`ds <= 0` では下側境界までの距離を返し、DDA の `tmx`、`tmy`、`tmz` 初期値を決定する。
    """
    if ds > 0:
      s = s - math.floor(s)
      return (1.0 - s) / ds
    else:
      s = s - math.floor(s)
      return s / (-ds)

  tmx = int_bound(origin.x / cell_size, d.x) if abs(d.x) > 1e-12 else 1e30
  tmy = int_bound(origin.y / cell_size, d.y) if abs(d.y) > 1e-12 else 1e30
  tmz = int_bound(origin.z / cell_size, d.z) if abs(d.z) > 1e-12 else 1e30

  tdx = (cell_size / abs(d.x)) if abs(d.x) > 1e-12 else 1e30
  tdy = (cell_size / abs(d.y)) if abs(d.y) > 1e-12 else 1e30
  tdz = (cell_size / abs(d.z)) if abs(d.z) > 1e-12 else 1e30

  t = 0.0
  enter_face = -1

  while t <= t_max:
    yield DDAHit(int(x), int(y), int(z), float(t), int(enter_face))

    if tmx < tmy and tmx < tmz:
      x += step_x
      t = tmx
      tmx += tdx
      enter_face = 1 if step_x > 0 else 0
    elif tmy < tmz:
      y += step_y
      t = tmy
      tmy += tdy
      enter_face = 3 if step_y > 0 else 2
    else:
      z += step_z
      t = tmz
      tmz += tdz
      enter_face = 5 if step_z > 0 else 4

File: native/ludoxel_mathematics/src/dda.rs
// SPDX-FileCopyrightText: 2026 Kento Konishi
// SPDX-License-Identifier: LicenseRef-All-Rights-Reserved

pub const MAX_TRAVERSAL_CELLS: usize = 262_144;

pub struct DdaHit {
  pub cell_x: i64,
  pub cell_y: i64,
  pub cell_z: i64,
  pub t: f64,
  pub enter_face: i32,
}

fn int_bound(s: f64, ds: f64) -> f64 {
  if ds > 0.0 {
    let frac = s - s.floor();
    (1.0 - frac) / ds
  } else {
    let frac = s - s.floor();
    frac / (-ds)
  }
}

pub fn dda_grid_traverse(ox: f64, oy: f64, oz: f64, dx: f64, dy: f64, dz: f64, t_max: f64, cell_size: f64) -> Result<Vec<DdaHit>, &'static str> {
  let mut hits = Vec::new();

  if dx.abs() < 1e-12 && dy.abs() < 1e-12 && dz.abs() < 1e-12 {
    return Ok(hits);
  }

  let mut x = (ox / cell_size).floor() as i64;
  let mut y = (oy / cell_size).floor() as i64;
  let mut z = (oz / cell_size).floor() as i64;

  let step_x: i64 = if dx > 0.0 { 1 } else { -1 };
  let step_y: i64 = if dy > 0.0 { 1 } else { -1 };
  let step_z: i64 = if dz > 0.0 { 1 } else { -1 };

  let mut tmx = if dx.abs() > 1e-12 { int_bound(ox / cell_size, dx) } else { 1e30 };
  let mut tmy = if dy.abs() > 1e-12 { int_bound(oy / cell_size, dy) } else { 1e30 };
  let mut tmz = if dz.abs() > 1e-12 { int_bound(oz / cell_size, dz) } else { 1e30 };

  let tdx = if dx.abs() > 1e-12 { cell_size / dx.abs() } else { 1e30 };
  let tdy = if dy.abs() > 1e-12 { cell_size / dy.abs() } else { 1e30 };
  let tdz = if dz.abs() > 1e-12 { cell_size / dz.abs() } else { 1e30 };

  let mut t = 0.0f64;
  let mut enter_face: i32 = -1;

  while t <= t_max {
    if hits.len() >= MAX_TRAVERSAL_CELLS {
      return Err("dda_grid_traverse exceeds the supported cell budget");
    }

    hits.push(DdaHit { cell_x: x, cell_y: y, cell_z: z, t, enter_face });

    if tmx < tmy && tmx < tmz {
      x += step_x;
      t = tmx;
      tmx += tdx;
      enter_face = if step_x > 0 { 1 } else { 0 };
    } else if tmy < tmz {
      y += step_y;
      t = tmy;
      tmy += tdy;
      enter_face = if step_y > 0 { 3 } else { 2 };
    } else {
      z += step_z;
      t = tmz;
      tmz += tdz;
      enter_face = if step_z > 0 { 5 } else { 4 };
    }
  }

  Ok(hits)
}
